x/blockdb/database.go

```diff
@@ -0,0 +1,1319 @@
+	combinedBufSize, err := safemath.Add(uint64(sizeOfBlockEntryHeader), uint64(len(block)))
+	if err != nil {
+		return fmt.Errorf("calculating combined buffer size would overflow for block %d: %w", bh.Height, err)
+	}
+	combinedBuf := make([]byte, combinedBufSize)
+	copy(combinedBuf, headerBytes)
+	copy(combinedBuf[sizeOfBlockEntryHeader:], block)
+	if _, err := dataFile.WriteAt(combinedBuf, int64(localOffset)); err != nil {
+		return fmt.Errorf("failed to write block to data file at offset %d: %w", offset, err)
+	}
+
+	if s.config.SyncToDisk {
+		if err := dataFile.Sync(); err != nil {
+			return fmt.Errorf("failed to sync data file after writing block %d: %w", bh.Height, err)
+		}
+	}
+	return nil
+}
+
+func (s *Database) updateBlockHeights(writtenBlockHeight BlockHeight) error {
+	s.updateBlockHeightsAtomically(func(current *blockHeights) *blockHeights {
+		if current == nil {
+			heights := &blockHeights{
+				maxBlockHeight:      writtenBlockHeight,
+				maxContiguousHeight: unsetHeight,
+			}
+			if writtenBlockHeight == s.header.MinHeight {
+				heights.maxContiguousHeight = writtenBlockHeight
+			}
+			return heights
+		}
+
+		updated := &blockHeights{
+			maxBlockHeight:      current.maxBlockHeight,
+			maxContiguousHeight: current.maxContiguousHeight,
+		}
+
+		// Update max block height if needed
+		if writtenBlockHeight > current.maxBlockHeight || current.maxBlockHeight == unsetHeight {
+			updated.maxBlockHeight = writtenBlockHeight
+		}
+
+		// Update max contiguous height logic
+		prevContiguousCandidate := uint64(unsetHeight)
+		if writtenBlockHeight > s.header.MinHeight {
+			prevContiguousCandidate = writtenBlockHeight - 1
+		}
+
+		if current.maxContiguousHeight == prevContiguousCandidate {
+			// We can extend the contiguous sequence. Try to extend it further
+			// by checking if the next height is also available, which would repair gaps in the sequence.
+			currentMax := writtenBlockHeight
+			for {
+				nextHeightToVerify, err := safemath.Add(currentMax, 1)
+				if err != nil {
+					s.log.Error("Failed to update block heights: overflow in height calculation",
+						zap.Uint64("currentMax", currentMax),
+						zap.Error(err),
+					)
+					break
+				}
+				// Check if we have indexed a block at the next height, which would extend our contiguous sequence
+				_, err = s.readIndexEntry(nextHeightToVerify)
+				if err != nil {
+					// If no block exists at this height, we've reached the end of our contiguous sequence
+					if errors.Is(err, ErrBlockNotFound) {
+						break
+					}
+
+					// log unexpected error
+					s.log.Error("Failed to update block heights: error reading index entry",
+						zap.Uint64("height", nextHeightToVerify),
+						zap.Error(err),
+					)
+					break
+				}
+				currentMax = nextHeightToVerify
+			}
+			updated.maxContiguousHeight = currentMax
+		}
+
+		return updated
+	})
+
+	// Check if we need to persist header on checkpoint interval
+	if writtenBlockHeight%s.config.CheckpointInterval == 0 {
```

Comment:
wondering if it makes more sense to persist not based on the block height but instead based on iteration? Something like:

```
	var checkpointIteration atomic.Uint64
	
	if checkpointIteration.Add(1) % s.config.CheckpointInterval == 0 {
		
	}
```

This way we are independent of the input and do it more uniformly.

Reply:
I'm personally fine if you want to add it in the future, maybe it makes sense to open an issue and assign it to you so it won't get forgotten, because I will sure forget this in the next morning